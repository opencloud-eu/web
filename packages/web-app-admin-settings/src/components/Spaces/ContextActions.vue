<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :action-options="{ resources: items }" />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, PropType, unref } from 'vue'
import { SpaceResource } from '@opencloud-eu/web-client'
import { ContextActionMenu } from '@opencloud-eu/web-pkg'

import {
  useSpaceActionsDelete,
  useSpaceActionsDisable,
  useSpaceActionsEditDescription,
  useSpaceActionsEditQuota,
  useSpaceActionsRename,
  useSpaceActionsRestore,
  useActionsShowDetails
} from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'ContextActions',
  components: { ContextActionMenu },
  props: {
    items: {
      type: Array as PropType<SpaceResource[]>,
      required: true
    }
  },
  setup(props) {
    const filterParams = computed(() => ({ resources: props.items }))

    const { actions: deleteActions } = useSpaceActionsDelete()
    const { actions: disableActions } = useSpaceActionsDisable()
    const { actions: editQuotaActions } = useSpaceActionsEditQuota()
    const { actions: editDescriptionActions } = useSpaceActionsEditDescription()
    const { actions: renameActions } = useSpaceActionsRename()
    const { actions: restoreActions } = useSpaceActionsRestore()
    const { actions: showDetailsActions } = useActionsShowDetails()

    const menuItemsPrimaryActions = computed(() =>
      [...unref(renameActions), ...unref(editDescriptionActions)].filter((item) =>
        item.isVisible(unref(filterParams))
      )
    )
    const menuItemsSecondaryActions = computed(() =>
      [
        ...unref(editQuotaActions),
        ...unref(disableActions),
        ...unref(restoreActions),
        ...unref(deleteActions)
      ].filter((item) => item.isVisible(unref(filterParams)))
    )
    const menuItemsSidebar = computed(() =>
      [...unref(showDetailsActions)].filter((item) => item.isVisible(unref(filterParams)))
    )

    const menuSections = computed(() => {
      const sections = []

      if (unref(menuItemsPrimaryActions).length) {
        sections.push({
          name: 'primaryActions',
          items: unref(menuItemsPrimaryActions)
        })
      }
      if (unref(menuItemsSecondaryActions).length) {
        sections.push({
          name: 'secondaryActions',
          items: unref(menuItemsSecondaryActions)
        })
      }
      if (unref(menuItemsSidebar).length) {
        sections.push({
          name: 'sidebar',
          items: unref(menuItemsSidebar)
        })
      }
      return sections
    })

    return {
      menuSections
    }
  }
})
</script>
