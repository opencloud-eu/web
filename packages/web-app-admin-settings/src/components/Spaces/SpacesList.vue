<template>
  <div id="space-list">
    <div class="space-filters flex justify-end flex-wrap items-end mx-4 mb-4">
      <oc-text-input
        id="spaces-filter"
        v-model="filterTerm"
        class="w-3xs"
        :label="$gettext('Search')"
        autocomplete="off"
      />
    </div>
    <oc-table
      class="settings-spaces-table"
      :sort-by="sortBy"
      :sort-dir="sortDir"
      :fields="fields"
      :data="paginatedItems"
      :highlighted="highlighted"
      :sticky="isSticky"
      :header-position="fileListHeaderY"
      :hover="true"
      padding-x="medium"
      @sort="handleSort"
      @contextmenu-clicked="showContextMenuOnRightClick"
      @highlight="fileClicked"
    >
      <template #selectHeader>
        <oc-checkbox
          size="large"
          :label="$gettext('Select all spaces')"
          :model-value="allSpacesSelected"
          :label-hidden="true"
          @update:model-value="
            allSpacesSelected ? unselectAllSpaces() : selectSpaces(paginatedItems)
          "
        />
      </template>
      <template #select="{ item }">
        <oc-checkbox
          size="large"
          :model-value="isSpaceSelected(item)"
          :option="item"
          :label="getSelectSpaceLabel(item)"
          :label-hidden="true"
          @update:model-value="selectSpace(item)"
          @click.stop="fileClicked([item, $event])"
        />
      </template>
      <template #icon>
        <oc-icon name="layout-grid" />
      </template>
      <template #name="{ item }">
        <span :data-test-space-name="item.name" v-text="item.name" />
      </template>
      <template #manager="{ item }">
        {{ getManagerNames(item) }}
      </template>
      <template #members="{ item }">
        {{ getMemberCount(item) }}
      </template>
      <template #totalQuota="{ item }"> {{ getTotalQuota(item) }}</template>
      <template #usedQuota="{ item }"> {{ getUsedQuota(item) }}</template>
      <template #remainingQuota="{ item }"> {{ getRemainingQuota(item) }}</template>
      <template #mdate="{ item }">
        <span
          v-oc-tooltip="formatDate(item.mdate)"
          tabindex="0"
          v-text="formatDateRelative(item.mdate)"
        />
      </template>
      <template #status="{ item }">
        <span class="flex items-center">
          <oc-icon
            v-oc-tooltip="item.disabled ? $gettext('Disabled') : $gettext('Enabled')"
            :name="item.disabled ? 'stop-circle' : 'play-circle'"
            size="small"
            fill-type="line"
          />
        </span>
      </template>
      <template #actions="{ item }">
        <div class="spaces-list-actions">
          <oc-button
            v-oc-tooltip="spaceDetailsLabel"
            :aria-label="spaceDetailsLabel"
            appearance="raw"
            class="ml-1 quick-action-button p-1 spaces-table-btn-details"
            @click.stop.prevent="showDetailsForSpace(item)"
          >
            <oc-icon name="information" fill-type="line" />
          </oc-button>
          <context-menu-quick-action
            ref="contextMenuButtonRef"
            :item="item"
            :title="item.name"
            class="spaces-table-btn-action-dropdown"
            @quick-action-clicked="showContextMenuOnBtnClick($event, item)"
          >
            <template #contextMenu>
              <slot name="contextMenu" :space="item" />
            </template>
          </context-menu-quick-action>
        </div>
      </template>
      <template #footer>
        <pagination :pages="totalPages" :current-page="currentPage" />
        <div class="text-center w-full my-2">
          <p class="text-role-on-surface-variant">{{ footerTextTotal }}</p>
          <p v-if="filterTerm" class="text-role-on-surface-variant">{{ footerTextFilter }}</p>
        </div>
      </template>
    </oc-table>
  </div>
</template>

<script setup lang="ts">
import {
  formatDateFromJSDate,
  formatRelativeDateFromJSDate,
  displayPositionedDropdown,
  formatFileSize,
  defaultFuseOptions,
  useKeyboardActions,
  ContextMenuBtnClickEventData,
  useIsTopBarSticky,
  useSharesStore
} from '@opencloud-eu/web-pkg'
import { ComponentPublicInstance, computed, nextTick, onMounted, ref, unref, watch } from 'vue'
import { getSpaceManagers, SpaceResource } from '@opencloud-eu/web-client'
import Mark from 'mark.js'
import Fuse from 'fuse.js'
import { useGettext } from 'vue3-gettext'
import { eventBus, SortDir } from '@opencloud-eu/web-pkg'
import { SideBarEventTopics } from '@opencloud-eu/web-pkg'
import { ContextMenuQuickAction } from '@opencloud-eu/web-pkg'
import {
  useFileListHeaderPosition,
  useRoute,
  useRouter,
  usePagination
} from '@opencloud-eu/web-pkg'
import { Pagination } from '@opencloud-eu/web-pkg'
import { perPageDefault, perPageStoragePrefix } from '../../defaults'
import { findIndex } from 'lodash-es'
import {
  useKeyboardTableMouseActions,
  useKeyboardTableNavigation
} from '../../composables/keyboardActions'
import { useSpaceSettingsStore } from '../../composables'
import { storeToRefs } from 'pinia'
import { FieldType } from '@opencloud-eu/design-system/helpers'

const router = useRouter()
const route = useRoute()
const language = useGettext()
const { $gettext } = language
const { isSticky } = useIsTopBarSticky()
const sharesStore = useSharesStore()

const { y: fileListHeaderY } = useFileListHeaderPosition('#admin-settings-app-bar')
const contextMenuButtonRef = ref(undefined)
const sortBy = ref('name')
const sortDir = ref(SortDir.Asc)
const filterTerm = ref('')
const markInstance = ref(undefined)

const lastSelectedSpaceIndex = ref(0)
const lastSelectedSpaceId = ref(null)

const spaceSettingsStore = useSpaceSettingsStore()
const { spaces, selectedSpaces } = storeToRefs(spaceSettingsStore)

const highlighted = computed(() => unref(selectedSpaces).map((s) => s.id))
const footerTextTotal = computed(() => {
  return $gettext('%{spaceCount} spaces in total', {
    spaceCount: unref(spaces).length.toString()
  })
})
const footerTextFilter = computed(() => {
  return $gettext('%{spaceCount} matching spaces', {
    spaceCount: unref(items).length.toString()
  })
})

const orderBy = (list: SpaceResource[], prop: string, desc: boolean) => {
  return [...list].sort((s1, s2) => {
    let a: string, b: string
    const numeric = ['totalQuota', 'usedQuota', 'remainingQuota'].includes(prop)

    switch (prop) {
      case 'members':
        a = getMemberCount(s1).toString()
        b = getMemberCount(s2).toString()
        break
      case 'totalQuota':
        a = (s1.spaceQuota.total || 0).toString()
        b = (s2.spaceQuota.total || 0).toString()
        break
      case 'usedQuota':
        a = (s1.spaceQuota.used || 0).toString()
        b = (s2.spaceQuota.used || 0).toString()
        break
      case 'remainingQuota':
        a = (s1.spaceQuota.remaining || 0).toString()
        b = (s2.spaceQuota.remaining || 0).toString()
        break
      case 'status':
        a = s1.disabled.toString()
        b = s2.disabled.toString()
        break
      default:
        a = s1[prop as keyof SpaceResource].toString() || ''
        b = s2[prop as keyof SpaceResource].toString() || ''
    }

    return desc
      ? b.localeCompare(a, undefined, { numeric })
      : a.localeCompare(b, undefined, { numeric })
  })
}
const items = computed(() =>
  orderBy(filter(unref(spaces), unref(filterTerm)), unref(sortBy), unref(sortDir) === SortDir.Desc)
)
const {
  items: paginatedItems,
  page: currentPage,
  total: totalPages
} = usePagination({ items, perPageDefault, perPageStoragePrefix })

const keyActions = useKeyboardActions()
useKeyboardTableNavigation(
  keyActions,
  paginatedItems,
  selectedSpaces,
  lastSelectedSpaceIndex,
  lastSelectedSpaceId
)
useKeyboardTableMouseActions(
  keyActions,
  paginatedItems,
  selectedSpaces,
  lastSelectedSpaceIndex,
  lastSelectedSpaceId
)

watch(currentPage, () => {
  unselectAllSpaces()
})

const allSpacesSelected = computed(() => {
  return unref(paginatedItems).length === unref(selectedSpaces).length
})

const handleSort = (event: { sortBy: string; sortDir: SortDir }) => {
  sortBy.value = event.sortBy
  sortDir.value = event.sortDir
}
const filter = (spaces: SpaceResource[], filterTerm: string) => {
  if (!(filterTerm || '').trim()) {
    return spaces
  }
  const searchEngine = new Fuse(spaces, { ...defaultFuseOptions, keys: ['name'] })
  return searchEngine.search(filterTerm).map((r) => r.item)
}
const isSpaceSelected = (space: SpaceResource) => {
  return unref(selectedSpaces).some((s) => s.id === space.id)
}

const fields = computed<FieldType[]>(() => [
  {
    name: 'select',
    title: '',
    type: 'slot',
    width: 'shrink',
    headerType: 'slot'
  },
  {
    name: 'icon',
    title: '',
    type: 'slot',
    width: 'shrink'
  },
  {
    name: 'name',
    title: $gettext('Name'),
    type: 'slot',
    sortable: true,
    tdClass: 'mark-element',
    width: 'expand'
  },
  {
    name: 'status',
    title: $gettext('Status'),
    type: 'slot',
    sortable: true
  },
  {
    name: 'manager',
    title: $gettext('Manager'),
    type: 'slot'
  },
  {
    name: 'members',
    title: $gettext('Members'),
    type: 'slot',
    sortable: true
  },
  {
    name: 'totalQuota',
    title: $gettext('Total quota'),
    type: 'slot',
    sortable: true
  },
  {
    name: 'usedQuota',
    title: $gettext('Used quota'),
    type: 'slot',
    sortable: true
  },
  {
    name: 'remainingQuota',
    title: $gettext('Remaining quota'),
    type: 'slot',
    sortable: true
  },
  {
    name: 'mdate',
    title: $gettext('Modified'),
    type: 'slot',
    sortable: true
  },

  {
    name: 'actions',
    title: $gettext('Actions'),
    sortable: false,
    type: 'slot',
    alignH: 'right'
  }
])

const getManagerNames = (space: SpaceResource) => {
  const allManagers = getSpaceManagers(space, sharesStore.graphRoles)
  if (!allManagers?.length) {
    return '-'
  }
  const managers = allManagers.length > 2 ? allManagers.slice(0, 2) : allManagers
  let managerStr = managers
    .map(({ grantedToV2 }) => (grantedToV2.user || grantedToV2.group).displayName)
    .join(', ')
  if (allManagers.length > 2) {
    managerStr += `... +${allManagers.length - 2}`
  }
  return managerStr
}
const formatDate = (date: string) => {
  return formatDateFromJSDate(new Date(date), language.current)
}
const formatDateRelative = (date: string) => {
  return formatRelativeDateFromJSDate(new Date(date), language.current)
}
const getTotalQuota = (space: SpaceResource) => {
  if (space.spaceQuota.total === 0) {
    return $gettext('Unrestricted')
  }

  return formatFileSize(space.spaceQuota.total, language.current)
}
const getUsedQuota = (space: SpaceResource) => {
  if (space.spaceQuota.used === undefined) {
    return '-'
  }
  return formatFileSize(space.spaceQuota.used, language.current)
}
const getRemainingQuota = (space: SpaceResource) => {
  if (space.spaceQuota.total === 0) {
    return $gettext('Unrestricted')
  }
  if (space.spaceQuota.remaining === undefined) {
    return '-'
  }
  return formatFileSize(space.spaceQuota.remaining, language.current)
}
const getMemberCount = (space: SpaceResource) => {
  return space.root.permissions?.length || 1
}

const getSelectSpaceLabel = (space: SpaceResource) => {
  return $gettext('Select %{ space }', { space: space.name }, true)
}

onMounted(() => {
  nextTick(() => {
    markInstance.value = new Mark('.mark-element')
  })
})

watch(filterTerm, async () => {
  await unref(router).push({ ...unref(route), query: { ...unref(route).query, page: '1' } })
})

watch([filterTerm, paginatedItems], () => {
  unref(markInstance)?.unmark()
  unref(markInstance)?.mark(unref(filterTerm), {
    element: 'span',
    className: 'mark-highlight'
  })
})

const fileClicked = (data: [SpaceResource, MouseEvent]) => {
  const resource = data[0]
  const eventData = data[1]
  const isCheckboxClicked = (eventData?.target as HTMLElement).getAttribute('type') === 'checkbox'

  const contextActionClicked =
    (eventData?.target as HTMLElement)?.closest('div')?.id === 'oc-files-context-menu'
  if (contextActionClicked) {
    return
  }
  if (!eventData.shiftKey && !eventData.metaKey && !eventData.ctrlKey) {
    eventBus.publish('app.files.shiftAnchor.reset')
  }
  if (eventData?.metaKey) {
    return eventBus.publish('app.resources.list.clicked.meta', resource)
  }
  if (eventData?.shiftKey) {
    return eventBus.publish('app.resources.list.clicked.shift', {
      resource,
      skipTargetSelection: isCheckboxClicked
    })
  }
  if (isCheckboxClicked) {
    return
  }

  unselectAllSpaces()
  selectSpace(resource)
}

const showContextMenuOnBtnClick = (data: ContextMenuBtnClickEventData, space: SpaceResource) => {
  const { dropdown, event } = data
  if (dropdown?.tippy === undefined) {
    return
  }
  if (!isSpaceSelected(space)) {
    spaceSettingsStore.setSelectedSpaces([space])
  }
  displayPositionedDropdown(dropdown.tippy, event, unref(contextMenuButtonRef))
}
const showContextMenuOnRightClick = (
  row: ComponentPublicInstance<unknown>,
  event: MouseEvent,
  space: SpaceResource
) => {
  event.preventDefault()
  const dropdown = row.$el.getElementsByClassName('spaces-table-btn-action-dropdown')[0]
  if (dropdown === undefined) {
    return
  }
  if (!isSpaceSelected(space)) {
    spaceSettingsStore.setSelectedSpaces([space])
  }
  displayPositionedDropdown(dropdown._tippy, event, unref(contextMenuButtonRef))
}

const spaceDetailsLabel = computed(() => {
  return $gettext('Show details')
})
const showDetailsForSpace = (space: SpaceResource) => {
  selectSpace(space)
  eventBus.publish(SideBarEventTopics.open)
}

const selectSpace = (selectedSpace: SpaceResource) => {
  lastSelectedSpaceIndex.value = findIndex(unref(spaces), (g) => g.id === selectedSpace.id)
  lastSelectedSpaceId.value = selectedSpace.id
  keyActions.resetSelectionCursor()

  const isSpaceSelected = unref(selectedSpaces).find((space) => space.id === selectedSpace.id)
  if (!isSpaceSelected) {
    return spaceSettingsStore.addSelectedSpace(selectedSpace)
  }

  spaceSettingsStore.setSelectedSpaces(
    unref(selectedSpaces).filter((space) => space.id !== selectedSpace.id)
  )
}

const unselectAllSpaces = () => {
  spaceSettingsStore.setSelectedSpaces([])
}

const selectSpaces = (spaces: SpaceResource[]) => {
  spaceSettingsStore.setSelectedSpaces(spaces)
}
</script>
<style>
@reference '@opencloud-eu/design-system/tailwind';

@layer utilities {
  .settings-spaces-table .oc-table-header-cell-actions,
  .settings-spaces-table .oc-table-data-cell-actions {
    @apply whitespace-nowrap;
  }

  /* Status, Members, Mdate: hidden by default, visible from md */
  .oc-table-header-cell-status,
  .oc-table-data-cell-status,
  .oc-table-header-cell-members,
  .oc-table-data-cell-members,
  .oc-table-header-cell-mdate,
  .oc-table-data-cell-mdate {
    @apply hidden md:table-cell;
  }

  /* Manager, TotalQuota, UsedQuota, RemainingQuota: hidden by default, visible from lg */
  .settings-spaces-table .oc-table-header-cell-manager,
  .settings-spaces-table .oc-table-data-cell-manager,
  .settings-spaces-table .oc-table-header-cell-totalQuota,
  .settings-spaces-table .oc-table-data-cell-totalQuota,
  .settings-spaces-table .oc-table-header-cell-usedQuota,
  .settings-spaces-table .oc-table-data-cell-usedQuota,
  .settings-spaces-table .oc-table-header-cell-remainingQuota,
  .settings-spaces-table .oc-table-data-cell-remainingQuota {
    @apply hidden lg:table-cell;
  }

  /* Squashed variant */
  .settings-spaces-table-squashed .oc-table-header-cell-manager,
  .settings-spaces-table-squashed .oc-table-data-cell-manager,
  .settings-spaces-table-squashed .oc-table-header-cell-status,
  .settings-spaces-table-squashed .oc-table-data-cell-status,
  .settings-spaces-table-squashed .oc-table-header-cell-members,
  .settings-spaces-table-squashed .oc-table-data-cell-members,
  .settings-spaces-table-squashed .oc-table-header-cell-totalQuota,
  .settings-spaces-table-squashed .oc-table-data-cell-totalQuota,
  .settings-spaces-table-squashed .oc-table-header-cell-usedQuota,
  .settings-spaces-table-squashed .oc-table-data-cell-usedQuota {
    @apply hidden;
  }

  /* RemainingQuota, Status, Members, Mdate visible from xl */
  .settings-spaces-table-squashed .oc-table-header-cell-remainingQuota,
  .settings-spaces-table-squashed .oc-table-data-cell-remainingQuota,
  .settings-spaces-table-squashed .oc-table-header-cell-status,
  .settings-spaces-table-squashed .oc-table-data-cell-status,
  .settings-spaces-table-squashed .oc-table-header-cell-members,
  .settings-spaces-table-squashed .oc-table-data-cell-members,
  .settings-spaces-table-squashed .oc-table-header-cell-mdate,
  .settings-spaces-table-squashed .oc-table-data-cell-mdate {
    @apply hidden xl:table-cell;
  }
}
</style>
