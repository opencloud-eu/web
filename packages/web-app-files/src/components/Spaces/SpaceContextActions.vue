<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :action-options="actionOptions" />
    <input
      id="space-image-upload-input"
      ref="spaceImageInput"
      type="file"
      name="file"
      multiple
      tabindex="-1"
      accept="image/jpeg, image/png"
      @change="showModalImageSpace"
    />
  </div>
</template>

<script lang="ts">
import {
  ContextActionMenu,
  FileActionOptions,
  isLocationSpacesActive,
  SpaceActionOptions,
  useFileActionsDownloadArchive,
  useFileActionsShowDetails,
  useRouter,
  useSpaceActionsDelete,
  useSpaceActionsDisable,
  useSpaceActionsDuplicate,
  useSpaceActionsEditDescription,
  useSpaceActionsEditQuota,
  useSpaceActionsEditReadmeContent,
  useSpaceActionsNavigateToTrash,
  useSpaceActionsRename,
  useSpaceActionsRestore,
  useSpaceActionsSetIcon,
  useSpaceActionsShowMembers
} from '@opencloud-eu/web-pkg'
import { useSpaceActionsUploadImage } from '../../composables'
import { computed, defineComponent, PropType, Ref, ref, toRef, unref, VNodeRef } from 'vue'

export default defineComponent({
  name: 'SpaceContextActions',
  components: { ContextActionMenu },
  props: {
    actionOptions: {
      type: Object as PropType<SpaceActionOptions>,
      required: true
    }
  },
  setup(props) {
    const router = useRouter()

    const actionOptions = toRef(props, 'actionOptions') as Ref<SpaceActionOptions>

    const { actions: deleteActions } = useSpaceActionsDelete()
    const { actions: disableActions } = useSpaceActionsDisable()
    const { actions: duplicateActions } = useSpaceActionsDuplicate()
    const { actions: editQuotaActions } = useSpaceActionsEditQuota()
    const { actions: editReadmeContentActions } = useSpaceActionsEditReadmeContent()
    const { actions: editDescriptionActions } = useSpaceActionsEditDescription()
    const { actions: setSpaceIconActions } = useSpaceActionsSetIcon()
    const { actions: renameActions } = useSpaceActionsRename()
    const { actions: restoreActions } = useSpaceActionsRestore()
    const { actions: showDetailsActions } = useFileActionsShowDetails()
    const { actions: showMembersActions } = useSpaceActionsShowMembers()
    const { actions: downloadArchiveActions } = useFileActionsDownloadArchive()
    const { actions: navigateToTrashActions } = useSpaceActionsNavigateToTrash()

    const spaceImageInput: VNodeRef = ref(null)
    const { actions: uploadImageActions, showModalImageSpace } = useSpaceActionsUploadImage({
      spaceImageInput
    })

    const menuItemsMembers = computed(() => {
      const fileHandlers = [...unref(showMembersActions), ...unref(downloadArchiveActions)]
      // HACK: downloadArchiveActions requires FileActionOptions but we have SpaceActionOptions
      return [...fileHandlers].filter((item) => item.isVisible(unref(actionOptions) as any))
    })

    const menuItemsPrimaryActions = computed(() => {
      const fileHandlers = [
        ...unref(renameActions),
        ...unref(duplicateActions),
        ...unref(editDescriptionActions),
        ...unref(uploadImageActions),
        ...unref(setSpaceIconActions)
      ]

      if (isLocationSpacesActive(router, 'files-spaces-generic')) {
        fileHandlers.splice(2, 0, ...unref(editReadmeContentActions))
      }
      return [...fileHandlers].filter((item) => item.isVisible(unref(actionOptions)))
    })

    const menuItemsSecondaryActions = computed(() => {
      const fileHandlers = [
        ...unref(editQuotaActions),
        ...unref(disableActions),
        ...unref(restoreActions),
        ...unref(navigateToTrashActions),
        ...unref(deleteActions)
      ]

      return [...fileHandlers].filter((item) => item.isVisible(unref(actionOptions)))
    })

    const menuItemsSidebar = computed(() => {
      const fileHandlers = [...unref(showDetailsActions)]
      return [...fileHandlers].filter((item) =>
        // HACK: showDetails provides FileAction[] but we have SpaceActionOptions, so we need to cast them to FileActionOptions
        item.isVisible(unref(actionOptions) as unknown as FileActionOptions)
      )
    })

    const menuSections = computed(() => {
      const sections = []
      if (unref(menuItemsMembers).length) {
        sections.push({
          name: 'members',
          items: unref(menuItemsMembers)
        })
      }
      if (unref(menuItemsPrimaryActions).length) {
        sections.push({
          name: 'primaryActions',
          items: unref(menuItemsPrimaryActions)
        })
      }
      if (unref(menuItemsSecondaryActions).length) {
        sections.push({
          name: 'secondaryActions',
          items: unref(menuItemsSecondaryActions)
        })
      }
      if (unref(menuItemsSidebar).length) {
        sections.push({
          name: 'sidebar',
          items: unref(menuItemsSidebar)
        })
      }

      return sections
    })

    return {
      menuSections,
      spaceImageInput,
      uploadImageActions,
      showModalImageSpace
    }
  }
})
</script>

<style lang="scss">
#space-image-upload-input {
  position: absolute;
  left: -99999px;
}
</style>
