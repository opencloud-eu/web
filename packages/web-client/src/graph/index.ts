import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
