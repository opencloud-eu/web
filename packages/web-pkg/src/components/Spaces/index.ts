export { default as QuotaModal } from './QuotaModal.vue'
export { default as SpaceImageModal } from './SpaceImageModal.vue'
