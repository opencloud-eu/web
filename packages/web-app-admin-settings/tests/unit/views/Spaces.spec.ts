import { SpaceResource } from '@opencloud-eu/web-client'
import { Graph } from '@opencloud-eu/web-client/graph'
import { mockDeep } from 'vitest-mock-extended'
import { ClientService, useAppDefaults } from '@opencloud-eu/web-pkg'
import {
  defaultComponentMocks,
  defaultPlugins,
  mount,
  useAppDefaultsMock
} from '@opencloud-eu/web-test-helpers'
import Spaces from '../../../src/views/Spaces.vue'

vi.mock('@opencloud-eu/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  queryItemAsString: vi.fn(),
  useAppDefaults: vi.fn(),
  useRouteQueryPersisted: vi.fn()
}))
vi.mocked(useAppDefaults).mockImplementation(() => useAppDefaultsMock({}))

const selectors = {
  loadingSpinnerStub: 'app-loading-spinner-stub',
  spacesListStub: 'spaces-list-stub',
  noContentMessageStub: 'no-content-message-stub',
  batchActionsStub: 'batch-actions-stub'
}

describe('Spaces view', () => {
  describe('loading states', () => {
    it('should show loading spinner if loading', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.find(selectors.loadingSpinnerStub).exists()).toBeTruthy()
    })
    it('should render spaces list after loading has been finished', async () => {
      const spaces = [{ id: '1', name: 'Some Space' }] as SpaceResource[]
      const { wrapper } = getWrapper({ spaces })
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.find(selectors.spacesListStub).exists()).toBeTruthy()
    })
  })
  it('should render no content message if no spaces found', async () => {
    const graph = mockDeep<Graph>()
    graph.drives.listAllDrives.mockResolvedValue([])
    const { wrapper } = getWrapper({ spaces: [] })
    await wrapper.vm.loadResourcesTask.last
    expect(wrapper.find(selectors.noContentMessageStub).exists()).toBeTruthy()
  })
  describe('batch actions', () => {
    it('do not display when no space selected', async () => {
      const { wrapper } = getWrapper()
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.find(selectors.batchActionsStub).exists()).toBeFalsy()
    })
    it('display when one space selected', async () => {
      const spaces = [{ id: '1', name: 'Some Space' }] as SpaceResource[]
      const { wrapper } = getWrapper({ spaces, selectedSpaces: spaces })
      await wrapper.vm.loadResourcesTask.last
      await wrapper.vm.$nextTick()
      expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
    })
    it('display when more than one space selected', async () => {
      const spaces = [
        { id: '1', name: 'Some Space' },
        { id: '1', name: 'Some other Space' }
      ] as SpaceResource[]
      const { wrapper } = getWrapper({ spaces, selectedSpaces: spaces })
      await wrapper.vm.loadResourcesTask.last
      await wrapper.vm.$nextTick()
      expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
    })
  })
})

function getWrapper({
  spaces = [
    {
      id: '1',
      name: 'space'
    } as SpaceResource
  ],
  selectedSpaces = []
}: { spaces?: SpaceResource[]; selectedSpaces?: SpaceResource[] } = {}) {
  const $clientService = mockDeep<ClientService>()
  $clientService.graphAuthenticated.drives.listAllDrives.mockResolvedValue(spaces)
  const mocks = {
    ...defaultComponentMocks(),
    $clientService
  }

  return {
    wrapper: mount(Spaces, {
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              spaceSettingsStore: {
                spaces,
                selectedSpaces
              }
            }
          })
        ],
        mocks,
        provide: mocks,
        stubs: {
          AppLoadingSpinner: true,
          NoContentMessage: true,
          SpacesList: true,
          OcBreadcrumb: true,
          BatchActions: true,
          ViewOptions: true
        }
      }
    })
  }
}
