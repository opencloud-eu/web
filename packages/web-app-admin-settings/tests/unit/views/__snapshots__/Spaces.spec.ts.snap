// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spaces view > loading states > should render spaces list after loading has been finished 1`] = `
"<div>
  <main class="oc-flex oc-height-1-1 app-content oc-width-1-1">
    <div id="admin-settings-wrapper" class="oc-width-expand oc-height-1-1 oc-position-relative">
      <div id="admin-settings-app-bar" class="oc-app-bar oc-py-s admin-settings-app-bar-sticky">
        <div class="admin-settings-app-bar-controls oc-flex oc-flex-between oc-flex-middle">
          <oc-breadcrumb-stub items="[object Object],[object Object]" id="admin-settings-breadcrumb" variation="default" contextmenupadding="medium" maxwidth="-1" truncationoffset="2" showcontextactions="false"></oc-breadcrumb-stub>
          <portal-target name="app.runtime.mobile.nav"></portal-target>
          <div class="oc-flex">
            <view-options-stub perpagestorageprefix="admin-settings" hashiddenfiles="false" hasfileextensions="false" haspagination="true" paginationoptions="20,50,100,250" perpagequeryname="items-per-page" perpagedefault="50" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
          </div>
        </div>
        <div class="admin-settings-app-bar-actions oc-flex oc-flex-middle oc-mt-xs">
          <!--v-if-->
        </div>
      </div>
      <div>
        <spaces-list-stub class=""></spaces-list-stub>
      </div>
    </div>
    <!--v-if-->
  </main>
</div>"
`;
